self-destruction with choices

peterbe opened the pull request:
Fixes #5862

I don't think you need to go through the same trouble I did to manually test this. But here's what I did:

1. First I log in as `peterbe` and click myself to http://wiki.mdn.localhost:8000/en-US/profiles/peterbe/delete
I pick the "Switch all my attributions for page changes to "Anonymous", and delete my account." option. 
<img width="775" alt="Screen Shot 2019-11-22 at 2 11 32 PM" src="https://user-images.githubusercontent.com/26739/69455244-fc90d780-0d35-11ea-837b-60afde9b136d.png">

When I'm done, I get signed out. I can sign in again but it's back into the same account but now with 0 revisions. 

2. Make my way back into the user delete page but now select the other option:
<img width="813" alt="Screen Shot 2019-11-22 at 3 05 03 PM" src="https://user-images.githubusercontent.com/26739/69456790-87270600-0d39-11ea-86ac-8ae636e72a64.png">

When I'm done, I get signed out.
In a separate browser I viewed my public profile before and after.

Before:
<img width="1143" alt="Screen Shot 2019-11-22 at 3 21 18 PM" src="https://user-images.githubusercontent.com/26739/69457781-d79f6300-0d3b-11ea-88c8-f2bdada23e3b.png">

After:
<img width="1144" alt="Screen Shot 2019-11-22 at 3 22 13 PM" src="https://user-images.githubusercontent.com/26739/69457797-e1c16180-0d3b-11ea-98d0-224b58fbe418.png">

peterbe commented:
I took the liberty to change the page title. Thanks @atopal !
<img width="868" alt="Screen Shot 2019-11-25 at 9 10 34 AM" src="https://user-images.githubusercontent.com/26739/69547402-87a2e500-0f63-11ea-9c68-3dd5917f0a3f.png">

escattone commented on kuma/users/tests/test_views.py:
```diff
@@ -1295,26 +1297,203 @@ def test_account_refresh_token_saved_next_login(self):
+            avatar_url='https://avatars0.githubusercontent.com/yada/yada',
+            html_url="https://github.com/{}".format(wiki_user.username)
+        )
+    )
```
You can use the `wiki_user_github_account` pytest fixture instead of creating this yourself.

escattone commented on kuma/users/tests/test_views.py:
```diff
@@ -1295,26 +1297,203 @@ def test_account_refresh_token_saved_next_login(self):
+        document=document,
+        creator=wiki_user,
+        save=True)
+    assert Revision.objects.filter(creator=wiki_user).exists()
```
All of this can be avoided by using the `create_revision` pytest fixture.

peterbe replied:
I don't understand that. The `create_revision` fixture is a function that returns an instance. It doesn't create anything.

escattone replied:
Oops, sorry, you could just use the `root_doc` fixture, which provides a doc and its revision.

escattone commented on kuma/users/tests/test_views.py:
```diff
@@ -1295,26 +1297,203 @@ def test_account_refresh_token_saved_next_login(self):
+            avatar_url='https://avatars0.githubusercontent.com/yada/yada',
+            html_url="https://github.com/{}".format(wiki_user.username)
+        )
+    )
```
Same here, you can use the `wiki_user_github_account` pytest fixture instead of creating this yourself.

escattone commented on kuma/users/tests/test_views.py:
```diff
@@ -1295,26 +1297,203 @@ def test_account_refresh_token_saved_next_login(self):
+        document=document,
+        creator=wiki_user,
+        save=True)
+    assert Revision.objects.filter(creator=wiki_user).exists()
```
Same here, you could use the `create_revision` pytest fixture instead of this chunk of code.

peterbe replied:
I don't undertstand that fixture. It's a fixture that returns an instance but doesn't create anything. 
However, `from kuma.wiki.tests import revision as create_revision` gets me a function that creates Revisions.

escattone replied:
Oops, sorry, you could just use the `root_doc` fixture, which provides a doc and its revision.

peterbe commented on kuma/users/jinja2/users/user_delete.html:
```diff
@@ -28,46 +28,21 @@ <h1>{{ title }}</h1>
-          Please file a bug to request closing your account, and ensure one of
-          these options is in the the description field:
+          {% trans count=revisions.count() %}
+          You have {{ count }} wiki revisions.
```
Both are extremely fast and extremely rare. I'll opt for clunkiness for now.

peterbe commented on kuma/users/views.py:
```diff
@@ -450,14 +455,87 @@ def user_edit(request, username):
+        request.session.clear()
+
+    def delete_user():
+        user.revisionakismetsubmission_set.all().delete()
```
Why keep them? The `AkismetSubmission` is ["used as a log of sorts"](https://github.com/mdn/kuma/blob/3ea95bb7d3c53a6d1a08eaf45b5f6b66b80c0e39/kuma/spam/models.py#L22) and if we replace the user with `anon`, what point is it to keep. All it will tell you is that 'ham' or 'spam' was submitted at {date} by some anonymous user. 

I'm just not convinced there's any value in keep the data. Granted, if it is a log, we also don't open the /var/log/someapp/things.log file and rewrite it.

escattone replied:
I'm thinking it's like removing an entry from a log when there's still value in the entry. We have the information that an Akismet "ham" or "spam" submission was made at a specific time/date on a specific document revision, and that information still seems valuable even though the sender is now anonymous.

peterbe replied:
Nevermind, I changed so that they're moved over to the anonymous user.

peterbe commented:
Wow! This was a lot more work than I had envisioned. Either way, I think I've addressed all your feedback now. 
Basically, it a LOT more careful with all the protected foreign key fields to the user before it does the `user.delete()` thing.

peterbe commented:
Also, bullet point removed:
<img width="709" alt="Screen Shot 2019-12-11 at 3 58 15 PM" src="https://user-images.githubusercontent.com/26739/70659952-1c2d6880-1c2f-11ea-9ebd-a2b989ae1d05.png">

peterbe commented:
I finally get it! As mentioned in Slack, the fixture `create_revision` was poorly named. I guess it should be called `root_doc_revision`. 
But I didn't want to correct that too. So I just switched to `root_doc` and used `root_doc.revisions.first()` to get my `revision` instance. 
I suppose we can tackle that fixture name at a later day.
